src/state/nonces/index.ts

```diff
@@ -16,15 +18,45 @@ type GetNonceArgs = {
 
 type UpdateNonceArgs = NonceData & GetNonceArgs;
 
-export async function getNextNonce({ address, chainId }: { address: string; chainId: ChainId }) {
+export async function getNextNonce({ address, chainId }: { address: string; chainId: ChainId }): Promise<number> {
   const { getNonce } = nonceStore.getState();
   const localNonceData = getNonce({ address, chainId });
   const localNonce = localNonceData?.currentNonce || 0;
-  const provider = getProvider({ chainId });
-  const txCountIncludingPending = await provider.getTransactionCount(address, 'pending');
-  if (!localNonce && !txCountIncludingPending) return 0;
-  const ret = Math.max(localNonce + 1, txCountIncludingPending);
-  return ret;
+  const provider = getBatchedProvider({ chainId });
+  const privateMempoolTimeout = chainsPrivateMempoolTimeout[chainId];
+
+  const pendingTxCountRequest = provider.getTransactionCount(address, 'pending');
+  const latestTxCountRequest = provider.getTransactionCount(address, 'latest');
+  const [pendingTxCountFromPublicRpc, latestTxCountFromPublicRpc] = await Promise.all([pendingTxCountRequest, latestTxCountRequest]);
+  const numPendingPublicTx = pendingTxCountFromPublicRpc - latestTxCountFromPublicRpc;
+  const numPendingLocalTx = Math.max(localNonce - latestTxCountFromPublicRpc, 0);
+  if (numPendingLocalTx === numPendingPublicTx) return pendingTxCountFromPublicRpc; // nothing in private mempool, proceed normally
+  if (numPendingLocalTx === 0 && numPendingPublicTx > 0) return latestTxCountFromPublicRpc; // catch up with public
+
+  const { pendingTransactions: storePendingTransactions } = pendingTransactionsStore.getState();
+  const pendingTransactions: RainbowTransaction[] = storePendingTransactions[address]?.filter(txn => txn.chainId === chainId) || [];
+
+  let nextNonce = localNonce + 1;
+  for (const pendingTx of pendingTransactions) {
+    if (!pendingTx.nonce || pendingTx.nonce < pendingTxCountFromPublicRpc) {
+      continue;
+    } else {
+      if (!pendingTx.timestamp) continue;
+      if (pendingTx.nonce === pendingTxCountFromPublicRpc) {
+        if (Date.now() - pendingTx.timestamp > privateMempoolTimeout) {
+          nextNonce = pendingTxCountFromPublicRpc;
+          break;
```

Comment:
This looks clean to me, but I think this might be a good spot to explain this reassignment with a small comment for anyone in the future with less context.